### ISSUE TITLE
IoU calculation 

### ISSUE THREAD

**Issue by ltb801** [qubvel/segmentation_models.pytorch#278]
Hi, I'm pretty new to this field and I want to ask about your way of calculating IoU score. From what I see in the functional.py code, the IoU score is calculated by adding the intersection of each class together and divide by the sum of the union of all class. Did I understand it correctly?

And I also found another way of calculating IoU score (in the attached image), which I took from this journal article: https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8569372

They calculate the IoU score for each class and average them together. I would like to ask which way of calculating IoU is the better one and how they are different. Thank you very much. 
![iou](https://user-images.githubusercontent.com/70506111/98894963-12287f80-24fa-11eb-8eda-3d9f63907ae5.png)

**Comment by Hexmagic** [comment]
> Hi, I'm pretty new to this field and I want to ask about your way of calculating IoU score. From what I see in the functional.py code, the IoU score is calculated by adding the intersection of each class together and divide by the sum of the union of all class. Did I understand it correctly?
> 
> And I also found another way of calculating IoU score (in the attached image), which I took from this journal article: https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8569372
> 
> They calculate the IoU score for each class and average them together. I would like to ask which way of calculating IoU is the better one and how they are different. Thank you very much.
> ![iou](https://user-images.githubusercontent.com/70506111/98894963-12287f80-24fa-11eb-8eda-3d9f63907ae5.png)
> ![image](https://user-images.githubusercontent.com/70506111/98895220-b8748500-24fa-11eb-9ea3-ba888ee8f7d1.png)

I think they are same,just use different format input. spm use   `BxCxHxW` input(one-hot) ,and this method use `BxHxW` input

**Comment by MCvin** [comment]
I'm also puzzled with the actual implementation of iou and f_score. It seems to me that the functions are for single class, and that it will give bad results when performing multi-class segmentation, especially if the classes are unbalanced.
Looking at the segmentation_models repo for Keras, it looked like they implemented a mean iou that deals with class imbalance:
https://github.com/qubvel/segmentation_models/blob/94f624b7029deb463c859efbd92fa26f512b52b8/segmentation_models/base/functional.py#L61
Is there something I am missing here ?

**Comment by JulienMaille** [comment]
@MCvin are you talking about the `class_weights` parameter?

**Comment by MCvin** [comment]
In part yes. In the Pytorch implementation the score is calculated as the sum of intersections divided by the sum of unions, which seems strange to me:
https://github.com/qubvel/segmentation_models.pytorch/blob/33dc95082ea76fd83af2a8ebd7a4ecbd549aecfa/segmentation_models_pytorch/utils/functional.py#L34-L36
whereas in the Keras implementation it seems more correct, i.e. the sum is given the axis parameter and will result in a tensor of intersections and unions. Only at the end when returning the score, an average is made over the classes and with the possibility of giving class_weights:
https://github.com/qubvel/segmentation_models/blob/94f624b7029deb463c859efbd92fa26f512b52b8/segmentation_models/base/functional.py#L92-L97

**Comment by ShengtianSang** [comment]
> I'm also puzzled with the actual implementation of iou and f_score. It seems to me that the functions are for single class, and that it will give bad results when performing multi-class segmentation, especially if the classes are unbalanced. Looking at the segmentation_models repo for Keras, it looked like they implemented a mean iou that deals with class imbalance: https://github.com/qubvel/segmentation_models/blob/94f624b7029deb463c859efbd92fa26f512b52b8/segmentation_models/base/functional.py#L61 Is there something I am missing here

> In part yes. In the Pytorch implementation the score is calculated as the sum of intersections divided by the sum of unions, which seems strange to me:
> 
> https://github.com/qubvel/segmentation_models.pytorch/blob/33dc95082ea76fd83af2a8ebd7a4ecbd549aecfa/segmentation_models_pytorch/utils/functional.py#L34-L36
> 
> 
> whereas in the Keras implementation it seems more correct, i.e. the sum is given the axis parameter and will result in a tensor of intersections and unions. Only at the end when returning the score, an average is made over the classes and with the possibility of giving class_weights:
> https://github.com/qubvel/segmentation_models/blob/94f624b7029deb463c859efbd92fa26f512b52b8/segmentation_models/base/functional.py#L92-L97

I'm also confused by the implementation of IoU of the SMP, especially the iou metric, I feel the IoU score of SMP is wrong, it could be used for training as a loss function (soft IoU loss), but it's not a correct metric to select or evaluate model's performance.

**Comment by JulienMaille** [comment]
Could one of you suggest a simple PR, so we understand what change you are proposing?
